docs/config.md
<!--
SPDX-FileCopyrightText: 2021 Markus Sommer

SPDX-License-Identifier: GPL-3.0-or-later
-->

# Daemon Configuration

Here, we'll go over the [sample config-file](https://github.com/dtn7/dtn7-go/blob/master/cmd/dtnd/configuration.toml) and explain the available options in-depth.

## [core]

### store

Path to the on-disk store used to persist bundles, may be either a relative or absolute path.

The store consist of a [badgerhold](https://github.com/timshannon/badgerhold) database for metadata, the bundle contents are written directly to disk.

### inspect-all-bundles

Inspect the Administrative Records from all Bundles, not only those addressed to this very node.
By doing so, this node can learn about a delivery and clear its store without being told directly.

### node-id

Public name of the node.
Will be set as the default endpoint ID for all CLAs, unless a different one has been specified.

According to the [standard](https://tools.ietf.org/html/draft-ietf-dtn-bpbis-31#section-4.2.5.1.1), your ID should be:

- `dtn:none`, if you are actually nothing
- `dtn://` + the actual name + `/` <- for a unicast enpoint
- `dtn://~` + the name of your multicast group + `/`  <- for something that is not a unicast enpoint (it does not neccessarily have to be a multicast group, but that's the only use case I can think off the top of my head)

You might also use the [other](https://tools.ietf.org/html/draft-ietf-dtn-bpbis-31#section-4.2.5.1.2) adressing scheme, but we can't guarantee that such behaviour won't lead to death and/or dismemberment.

## [logging]

## [discovery]

## [agents]

## [[listen]]

## [[peer]]

## [routing]

Select the routing algorithm, you can choose on from the list `["epidemic", "spray", "binary_spray", "dtlsr", "prophet", "sensor-mule"]`

### epidemic

*Epidemic Routing* is the simplest delay-tolerant routing algorithm.
All bundles are always sent to all peers, which gives the best delivery probability, but also the highest overhead.
Note that `dtnd` keeps track of peers who have already been sent a specific bundle, so each bundle should only be forwarded to each peer once.

### spray

The *Spray & Wait* routing algorithm.

### binary_spray

The *Binary Spray & Wait* routing algorithm.

### dtlsr

The *Delay-Tolerant Link-State Routing* algorithm.

### prophet

The *PRoPHET* routing algorithm.

### sensor-mule

For dtn over equine carriers.


CHANGELOG.md
<!--
SPDX-FileCopyrightText: 2020, 2021 Markus Sommer
SPDX-FileCopyrightText: 2020, 2021 Alvar Penning

SPDX-License-Identifier: GPL-3.0-or-later
-->

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][keep-a-changelog], and this
project adheres to [Semantic Versioning][semantic-versioning].

<!--
Please keep the text width at 72 chars for easy copying into git tags.


Types of changes:

- Added       for new features.
- Changed     for changes in existing functionality.
- Deprecated  for soon-to-be removed features.
- Removed     for now removed features.
- Fixed       for any bug fixes.
- Security    in case of vulnerabilities
-->

## [Unreleased]
### Added
- Connect Administrative Records with Bundle and BundleBuilder.
- Make PingAgent configurable as a dtnd Agent.
- Simple "dtn-tool ping" command to interact with a PingAgent.
- Reassembly support for fragmented Bundles in the Store.

### Changed
- Structural refactoring:
    - Move all Go packages into the pkg directory.
    - Renamed bundle package to bpv7.
      This new name better describes that it is about bpbis and not just
      bundles. Furthermore, this allows to use the obvious variable name
      "bundle" for a bpv7.Bundle.
    - Unlink discovery from Core.
    - Move custom routing-extension block to the bpv7 package,
      where they shall reside with their brethren.
- Bump draft-ietf-dtn-bpbis version from 26 to 31.
    - Restrict dtn URI node name to `[\w-._]+`, _ietf-dtn-bpbis-27_.
    - "Block unintelligible" status report is replaced by new "Block
      unsupported" status report code for non processable blocks.
- Bump draft-ietf-dtn-tcpclv4 version from 21 to 23.
- Set Linux-specific socket options for a MTCP Client's connection to
  detect an abrupt connection loss.

### Fixed
- Include nil-check for EndpointID's internal representation.
- Close a MTCP Client after sending a keepalive failed.
- Enable parsing of primary blocks without a CRC value. When creating,
  an attempt is made to enforce a CRC, since this is necessary in the
  absence of an integrity block (BPSec).
- Exclude the peer discovery Manager's function field from the
  JSONFormatter used by logrus. Otherwise, the struct cannot be encoded.


## [0.9.0] - 2020-10-08
### Added
- Specific routing algorithm for data mules in sensor networks.
- Additional contrib files to used dtnd with systemd and ufw.
- GitHub Action to perform multiple Go Race Detector tests at night.
- Add WebSocket-based variant of TCPCLv4.

### Changed
- An invalid EndpointID struct is interpreted as dtn:none.
- Compare EndpointIDs based on both scheme and authority part.
- Allow any bundle.ExtensionBlock to implement a json.Marshaler for a
  more human readable output. This was done for the bpbis blocks.
- bundle.ExtensionBlock has a "BlockTypeName" next to its type code for
  a human readable output.
- Renamed cla/tcpcl package to cla/tcpclv4.
- Renamed tcpcl to tcpclv4 in dtnd's configuration.
- Bump draft-ietf-dtn-tcpclv4 version from 14 to 21.
- API change on cla.Convergence and cla.ConvergenceSender.
- Changed CLAType numbers to allow grouping similar Convergence Layers.

### Fixed
- Check if WebSocket Agent's channel is closed to mitigate dtn-tool from
  crashing.


## [0.8.0] - 2020-08-05
### Added
- REUSE compliance and a new copyright header generation script.
- Sort Bundle's CanonicalBlocks on creation and block modification.
- Custom SignatureBlock for cryptographic ed25519 Bundle signatures.
- dtnd supports attaching a SignatureBlock for Administrative Records.

### Changed
- Replaced TravisCI with GitHub Actions.
- List all Extension Block type codes in `bundle/extension_block.go`.
- Fragmentation tries to copy the original CRC type.
- BundleBuilder sets flags by default:
    - Request delivery status bundle control flag.
    - Replicate block control flag for Bundle Age Block, Hop Count
      Block, and Previous Node Block to ease Bundle fragmentation.

### Removed
- Drop compatibility with Go versions below 1.13.

### Fixed
- Ensure only Payload Blocks are allowed to get Block Number 1 when
  adding Extension Blocks to an empty Bundle.
- PrimaryBlock: always overwrite CRC, don't rely on cached values
- dtn-tool exchange: do not resend received Bundles


## [0.7.1] - 2020-08-03
### Fixed
- Restoring compatibility with Go 1.11.

### Deprecated
- Compatibility with Go 1.11 will soon be dropped. Expect the 0.7.x
  releases the last ones to support this now almost two year old Go
  version. Sorry.


## [0.7.0] - 2020-08-03
### Added
- `AdministrativeRecordManager` to allow more dynamic Administrative
  Records.
- `EndpointID` gets a singleton property, _ietf-dtn-bpbis-26_.
- Make repository more friendly for new contributors by
    - GitHub Issue template,
    - `CHANGELOG.md` file, and
    - Contributing section in `README.md`.

### Changed
- Enforce strict `dtn` URI scheme based on the ABNF, like this
  `dtn://NODE-NAME/OPTIONAL-VARIOUS-CHARS`.
- Allow peer discovery to work with multiple Endpoint IDs.
- CLA management is performed by the CLA Manager.
- Time is normed to milliseconds, _ietf-dtn-bpbis-26_:
    - DTN Time: milliseconds instead of seconds
    - Primary Block's lifetime: milliseconds instead of microseconds
    - Bundle Age Block: milliseconds instead of microseconds
- Bump draft-ietf-dtn-bpbis version from 24 to 26.

### Fixed
- `BundleBuilder` sorts CanonicalBlocks based on their block number.


## [0.6.1] - 2020-04-16
> _This release was created before adapting the
> [Keep a Changelog][keep-a-changelog] format._

Normalize RestAgent's error response field

Previously, the _error_ reporting in the JSON objects were differerent
for different methods. This makes a programming library unnecessarily
complicated.

Furthermore, the bundle/arecord package was moved into bundle. This was
actually only supposed to happen in the next main release.


## [0.6.0] - 2020-04-16

cmd/dtnd/configuration.go
// SPDX-FileCopyrightText: 2019, 2020 Markus Sommer
// SPDX-FileCopyrightText: 2019, 2020, 2021 Alvar Penning
//
// SPDX-License-Identifier: GPL-3.0-or-later

package main

import (
	"crypto/ed25519"
	"encoding/hex"
	"fmt"
	"net"
	"net/http"
	"strconv"
	"time"

	"github.com/gorilla/mux"
	log "github.com/sirupsen/logrus"

	"github.com/BurntSushi/toml"

	"github.com/dtn7/dtn7-go/pkg/agent"
	"github.com/dtn7/dtn7-go/pkg/bpv7"
	"github.com/dtn7/dtn7-go/pkg/cla"
	"github.com/dtn7/dtn7-go/pkg/cla/bbc"
	"github.com/dtn7/dtn7-go/pkg/cla/mtcp"
	"github.com/dtn7/dtn7-go/pkg/cla/tcpclv4"
	"github.com/dtn7/dtn7-go/pkg/discovery"
	"github.com/dtn7/dtn7-go/pkg/routing"
)

// tomlConfig describes the TOML-configuration.
type tomlConfig struct {
	Core      coreConf
	Logging   logConf
	Discovery discoveryConf
	Agents    agentsConfig
	Listen    []convergenceConf
	Peer      []convergenceConf
	Routing   routing.RoutingConf
}

// coreConf describes the Core-configuration block.
type coreConf struct {
	Store             string
	InspectAllBundles bool   `toml:"inspect-all-bundles"`
	NodeId            string `toml:"node-id"`
	SignPriv          string `toml:"signature-private"`
}

// logConf describes the Logging-configuration block.
type logConf struct {
	Level        string
	ReportCaller bool `toml:"report-caller"`
	Format       string
}

// discoveryConf describes the Discovery-configuration block.
type discoveryConf struct {
	IPv4     bool
	IPv6     bool
	Interval uint
}

// agentsConfig describes the ApplicationAgents/Agent-configuration block.
type agentsConfig struct {
	Ping      string
	Webserver agentsWebserverConfig
}

// agentsWebserverConfig describes the nested "Webserver" configuration for agents.
type agentsWebserverConfig struct {
	Address   string
	Websocket bool
	Rest      bool
}

// convergenceConf describes the Convergence-configuration block, used for
// "listen" and "peer".
type convergenceConf struct {
	Node     string
	Protocol string
	Endpoint string
}

func parseListenPort(endpoint string) (port int, err error) {
	var portStr string
	_, portStr, err = net.SplitHostPort(endpoint)
	if err != nil {
		return
	}
	port, err = strconv.Atoi(portStr)
	return
}

// parseListen inspects a "listen" convergenceConf and returns a Convergable.
func parseListen(conv convergenceConf, nodeId bpv7.EndpointID) (cla.Convergable, bpv7.EndpointID, cla.CLAType, discovery.Announcement, error) {
	log.WithFields(log.Fields{
		"EndpointID": conv.Node,
		"Endpoint":   conv.Endpoint,
		"Protocol":   conv.Protocol,
	}).Debug("Initialising convergence adaptor")

	// if the user has configured an EndpointID for this convergence adaptor
	if conv.Node != "" {
		parsedId, err := bpv7.NewEndpointID(conv.Node)
		if err != nil {
			return nil, nodeId, 0, discovery.Announcement{}, err
		} else {
			log.WithFields(log.Fields{
				"listener ID": conv.Node,
			}).Debug("Using alternative configured endpoint id for listener")
			nodeId = parsedId
		}
	}

	switch conv.Protocol {
	case "bbc":
		conn, err := bbc.NewBundleBroadcastingConnector(conv.Endpoint, true)
		return conn, nodeId, cla.BBC, discovery.Announcement{}, err

	case "mtcp":
		portInt, err := parseListenPort(conv.Endpoint)
		if err != nil {
			return nil, nodeId, cla.MTCP, discovery.Announcement{}, err
		}

		msg := discovery.Announcement{
			Type:     cla.MTCP,
			Endpoint: nodeId,
			Port:     uint(portInt),
		}

		return mtcp.NewMTCPServer(conv.Endpoint, nodeId, true), nodeId, cla.MTCP, msg, nil

	case "tcpclv4":
		portInt, err := parseListenPort(conv.Endpoint)
		if err != nil {
			return nil, nodeId, cla.TCPCLv4, discovery.Announcement{}, err
		}

		listener := tcpclv4.ListenTCP(conv.Endpoint, nodeId)

		msg := discovery.Announcement{
			Type:     cla.TCPCLv4,
			Endpoint: nodeId,
			Port:     uint(portInt),
		}

		return listener, nodeId, cla.TCPCLv4, msg, nil

	case "tcpclv4-ws":
		listener := tcpclv4.ListenWebSocket(nodeId)

		httpMux := http.NewServeMux()
		httpMux.Handle("/tcpclv4", listener)
		httpServer := &http.Server{
			Addr:    conv.Endpoint,
			Handler: httpMux,
		}

		errChan := make(chan error)
		go func() { errChan <- httpServer.ListenAndServe() }()

		select {
		case err := <-errChan:
			return nil, nodeId, cla.TCPCLv4WebSocket, discovery.Announcement{}, err

		case <-time.After(100 * time.Millisecond):
			return listener, nodeId, cla.TCPCLv4WebSocket, discovery.Announcement{}, nil
		}

	default:
		return nil, nodeId, 0, discovery.Announcement{}, fmt.Errorf("unknown listen.protocol \"%s\"", conv.Protocol)
	}
}

func parsePeer(conv convergenceConf, nodeId bpv7.EndpointID) (cla.ConvergenceSender, error) {

	switch conv.Protocol {
	case "mtcp":
		if endpointID, err := bpv7.NewEndpointID(conv.Node); err != nil {
			return nil, err
		} else {
			return mtcp.NewMTCPClient(conv.Endpoint, endpointID, true), nil
		}

	case "tcpclv4":
		return tcpclv4.DialTCP(conv.Endpoint, nodeId, true), nil

	case "tcpclv4-ws":
		return tcpclv4.DialWebSocket(conv.Endpoint, nodeId, true), nil

	default:
		return nil, fmt.Errorf("unknown peer.protocol \"%s\"", conv.Protocol)
	}
}

// parseAgents for the ApplicationAgents.
func parseAgents(conf agentsConfig) (agents []agent.ApplicationAgent, err error) {
	if conf.Ping != "" {
		if pingEid, pingEidErr := bpv7.NewEndpointID(conf.Ping); pingEidErr != nil {
			err = pingEidErr
			return
		} else {
			agents = append(agents, agent.NewPing(pingEid))
		}
	}

	if (conf.Webserver != agentsWebserverConfig{}) {
		if !conf.Webserver.Websocket && !conf.Webserver.Rest {
			err = fmt.Errorf("webserver agent needs at least one of Websocket or REST")
			return
		}

		r := mux.NewRouter()

		if conf.Webserver.Websocket {
			ws := agent.NewWebSocketAgent()
			r.HandleFunc("/ws", ws.ServeHTTP)

			agents = append(agents, ws)
		}

		if conf.Webserver.Rest {
			restRouter := r.PathPrefix("/rest").Subrouter()
			ra := agent.NewRestAgent(restRouter)

			agents = append(agents, ra)
		}

		httpServer := &http.Server{
			Addr:    conf.Webserver.Address,
			Handler: r,
		}

		errChan := make(chan error)
		go func() { errChan <- httpServer.ListenAndServe() }()

		select {
		case err = <-errChan:
			return

		case <-time.After(100 * time.Millisecond):
			break
		}
	}

	return
}

// parseCore creates the Core based on the given TOML configuration.
func parseCore(filename string) (c *routing.Core, ds *discovery.Manager, err error) {
	var conf tomlConfig
	if _, err = toml.DecodeFile(filename, &conf); err != nil {
		return
	}

	// Logging
	if conf.Logging.Level != "" {
		if lvl, err := log.ParseLevel(conf.Logging.Level); err != nil {
			log.WithFields(log.Fields{
				"level":    conf.Logging.Level,
				"error":    err,
				"provided": "panic,fatal,error,warn,info,debug,trace",
			}).Warn("Failed to set log level. Please select one of the provided ones")
		} else {
			log.SetLevel(lvl)
		}
	}

	log.SetReportCaller(conf.Logging.ReportCaller)

	switch conf.Logging.Format {
	case "", "text":
		log.SetFormatter(&log.TextFormatter{
			FullTimestamp:   true,
			TimestampFormat: "15:04:05.000",
		})

	case "json":
		log.SetFormatter(&log.JSONFormatter{
			TimestampFormat: time.RFC3339Nano,
		})

	default:
		log.Warn("Unknown logging format")
	}

	var discoveryMsgs []discovery.Announcement

	// Core
	if conf.Core.Store == "" {
		err = fmt.Errorf("routing.store is empty")
		return
	}

	log.WithFields(log.Fields{
		"routing": conf.Routing.Algorithm,
	}).Debug("Selected routing algorithm")

	nodeId, nodeErr := bpv7.NewEndpointID(conf.Core.NodeId)
	if nodeErr != nil {
		err = nodeErr
		return
	}

	var signPriv ed25519.PrivateKey = nil
	if conf.Core.SignPriv != "" {
		if signPriv, err = hex.DecodeString(conf.Core.SignPriv); err != nil {
			return
		}
	}

	if c, err = routing.NewCore(conf.Core.Store, nodeId, conf.Core.InspectAllBundles, conf.Routing, signPriv); err != nil {
		return
	}

	// Agents
	if conf.Agents != (agentsConfig{}) {
		if appAgents, appErr := parseAgents(conf.Agents); appErr != nil {
			err = appErr
			return
		} else {
			for _, appAgent := range appAgents {
				c.RegisterApplicationAgent(appAgent)
			}
		}
	}

	// Listen/ConvergenceReceiver
	for _, conv := range conf.Listen {
		if convRec, eid, claType, discoMsg, lErr := parseListen(conv, c.NodeId); lErr != nil {
			err = lErr
			return
		} else {
			c.RegisterCLA(convRec, claType, eid)
			if discoMsg != (discovery.Announcement{}) {
				discoveryMsgs = append(discoveryMsgs, discoMsg)
			}
		}
	}

	// Peer/ConvergenceSender
	for _, conv := range conf.Peer {
		convRec, err := parsePeer(conv, c.NodeId)
		if err != nil {
			log.WithFields(log.Fields{
				"peer":  conv.Endpoint,
				"error": err,
			}).Warn("Failed to establish a connection to a peer")
			continue
		}

		c.RegisterConvergable(convRec)
	}

	// Discovery
	if conf.Discovery.IPv4 || conf.Discovery.IPv6 {
		if conf.Discovery.Interval == 0 {
			conf.Discovery.Interval = 10
		}

		ds, err = discovery.NewManager(
			c.NodeId, c.RegisterConvergable, discoveryMsgs,
			time.Duration(conf.Discovery.Interval)*time.Second, conf.Discovery.IPv4, conf.Discovery.IPv6)
		if err != nil {
			return
		}
	}

	return
}
